### ISSUE TITLE
Bundling with webpack VS Lerna boostraping.

### ISSUE THREAD

**Issue by migrachev** [lerna/lerna#1049]
Hello I am having problems making lerna and webpack work together in peice.

**[Assumptions]**
- Lets imagine we have a _lerna-repo_ with several packages - _package1_, _package2_, ... ,_packageN_.
- Now imagine that _package1_ depends on _package2_ and _package2_ on the other hand depends on another _external-package_.
- Note that _package1_ does not depend on _external-package_ since it does not use it (import/require it) in any of its (own) source code.
- Now imagine that _package1_ needs to build itself to bundle.js via webpack (i believe which bundler is used here does not matter). This might be needed for deployment on a test server or may be just the way the package wants to be distributed, it shouldn't matter.

**[The problem]**
- As a result of _lerna bootstrap_  in _lerna-repo_, in lerna-repo/packages/package1/node_modules you will have _package2_ but you will not have _external-package_. Note that this will not be true if we forget about lerna for a second and simply _npm install_ package1's dependencies. Hence for some reason the result of _lerna bootstrap_ may be different from the result of npm install for your packages.
Having that in mind you will find webpack failing to bundle your code because of missing dependencies. And in fact they actually are missing in node_modules.

**[My assumptions]**
- In order to work around this issue we can couple the code in _package1_ and/or the webpack configuration with the fact that the package is within lerna-repo. However I strongly believe that such coupling is wrong and not intended by the lerna idea. After all this package may tomorrow migrate to some other package managing awesomeness.
- On the other hand the very idea that _lerna bootstrap_ puts an actual _link_ in your packages' node_modules (instead of saving that metadata somewhere else) makes me believe that lerna wants to simulate _npm install_ in your packages, allowing them to not concern that they are currently located in a lerna-repo.

**[My suggestion/question]**
- Do you find it possible to enhance _lerna bootstrap_ to more precisely simulate _npm install_ and add the correct symlinks for all the packages that _npm install_ would produce? I know it sounds way easier than it actually is, I am asking just about the idea for now.

**Comment by evocateur** [comment]
Do you have an example where this webpack bundling is failing to include a transitive dependency form a lerna-linked (aka "sibling") dependency?

I use webpack to create bundles from lerna packages daily, and I haven't had any issues like this.

Granted, I also explicitly `--hoist` everything to the root of the repo (adding to the `dependencies` of the root package.json), so most of my `packages/*/node_modules` directories are empty except for relative sibling symlinks. This isn't terribly common, despite being very handy (which I have intended to streamline for awhile, but haven't found the time...)

**Comment by migrachev** [comment]
At the moment I cannot give you such and example due to code privacy, but I will come up with one and link it.

**Comment by migrachev** [comment]
After attempting to reproduce the problem with minimum configuration, it turned out that webpack's resolving behavior is configurable and the reason why my described case didn't work was because I specifically say to webpack to resolve _package1_ node_modules. However when I allowed webpack to respect nested node_modules, I stumbled on another problem, so here is why I will try to give more detailed info here in case someone has similar problem.

[The problem]
The problem that I specified above was caused by custom webpack configuration. By default webpack should not have problems resolving the node_modules of your node_modules. What I was doing was:
```js
resolve: {
        modules: [
            path.join(__dirname, "node_modules")
        ]
}
```
With this configuration webpack will respect only __dirname/node_modules when resolving dependencies and (in the upper mentioned example) dependencies of our dependencies will not be properly resolved. Or the least to say they will be strangely resolved.
```js
resolve: {
        modules: [
            "node_modules"
        ]
}
```
This is the configuration that will allow webpack to respect nested node_modules. However in my case that end up in another problem.

[The New Problem]
My packages are using React and React have that restriction that only one instance of React must be loaded by your javascript. The webpack bundle.js was violating that. The reason for that is because my packages depend on React and that is quite normal because they use it. So when webpack transpile and bundle my code, it will load/require React twice.
To solve this problem I use alias for react (I think there are other solutions as well)
```js
resolve: {
        modules: ["node_modules"],
        alias: {
            react: path.resolve(__dirname, "node_modules/react")
        },
        extensions: [".js", ".css"]
    }
```
